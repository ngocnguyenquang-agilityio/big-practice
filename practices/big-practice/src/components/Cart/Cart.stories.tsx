// Types
import type { Meta, StoryObj } from "@storybook/react";

// Components
import Cart from ".";

const meta = {
  title: 'Components/Cart',
  component: Cart,
} satisfies Meta<typeof Cart>;

export default meta;
type Story = StoryObj<typeof meta>

export const Default: Story = {
  render: () => <Cart />
}
