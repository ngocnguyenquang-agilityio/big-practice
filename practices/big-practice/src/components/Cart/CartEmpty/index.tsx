// Components
import { CartIcon } from '@assets/icons';

const CartEmpty = () => {
  return (
    <div className='mt-20 flex w-full flex-col items-center justify-center overflow-hidden'>
      <CartIcon width={64} height={64} />
      <p className='mt-6 text-center text-2xl font-bold'>Your cart is empty.</p>
    </div>
  );
};

export default CartEmpty;
