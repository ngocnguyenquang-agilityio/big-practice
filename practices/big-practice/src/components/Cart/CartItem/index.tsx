// Icons
import removeIcon from '@assets/removeIcon.svg';

// Types
import { ICartItem } from '@interfaces';

// Components
import QuantityActionButton from '@components/QuantityActionButton';
import { Icon } from '@components/Icon';

const CartItem = ({ product, quantity = 1, handleRemoveItemFromCart, handleDecrease, handleIncrease }: ICartItem) => {
  return (
    <div className='relative flex w-full flex-row justify-between px-1 py-4'>
      <div className='absolute z-40 -mt-2 ml-[55px]'>
        <button
          className='ease flex h-[17px] w-[17px] items-center justify-center rounded-full bg-neutral-500 transition-all duration-200'
          onClick={handleRemoveItemFromCart}
        >
          <Icon
            svg={removeIcon}
            name='remove-icon'
          />
        </button>
      </div>
      <a
        className='z-30 flex flex-row space-x-4'
        href=''
      >
        <div className='relative h-16 w-16 cursor-pointer overflow-hidden rounded-md border border-neutral-700 bg-neutral-900 hover:bg-neutral-800'>
          <img
            className='h-full w-full object-cover'
            loading='lazy'
            src={product.thumbnail}
          />
        </div>
        <div className='flex flex-1 flex-col text-base'>
          <span className='leading-tight'>{product.title}</span>
          <p className='text-sm text-neutral-400'>Black / 7 x 9 inch</p>
        </div>
      </a>
      <div className='flex h-16 flex-col justify-between'>
        <p className='flex justify-end space-y-2 text-right text-sm'>
          {product.price}
          <span className='ml-1 inline'>USD</span>
        </p>
        <QuantityActionButton
          quantity={quantity}
          handleDecrease={handleDecrease}
          handleIncrease={handleIncrease}
        />
      </div>
    </div>
  );
};

export default CartItem;
