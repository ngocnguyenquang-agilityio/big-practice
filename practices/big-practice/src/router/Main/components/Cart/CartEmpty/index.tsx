const CartEmpty = () => {
  return (
    <div className='mt-20 flex w-full flex-col items-center justify-center overflow-hidden'>
      <svg
        name='cart-icon'
        xmlns='http://www.w3.org/2000/svg'
        width={64}
        height={64}
        viewBox='0 0 24 24'
        fill='none'
        stroke='currentColor'
        strokeWidth='2'
        strokeLinecap='round'
        strokeLinejoin='round'
        className='lucide lucide-shopping-cart'
      >
        <circle
          cx='8'
          cy='21'
          r='1'
        />
        <circle
          cx='19'
          cy='21'
          r='1'
        />
        <path d='M2.05 2.05h2l2.66 12.42a2 2 0 0 0 2 1.58h9.78a2 2 0 0 0 1.95-1.57l1.65-7.43H5.12' />
      </svg>
      <p className='mt-6 text-center text-2xl font-bold'>Your cart is empty.</p>
    </div>
  );
};

export default CartEmpty;
