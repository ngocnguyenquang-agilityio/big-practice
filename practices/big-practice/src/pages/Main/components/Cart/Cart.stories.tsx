// Types
import type { Meta, StoryObj } from '@storybook/react';

// Components
import Cart from '.';

const meta = {
  title: 'Components/Cart',
  component: Cart,
} satisfies Meta<typeof Cart>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    cart: [
      {
        id: 1,
        title: 'Product Detail',
        price: 20,
        thumbnail:
          'https://demo.vercel.store/_next/image?url=https%3A%2F%2Fcdn.shopify.com%2Fs%2Ffiles%2F1%2F0754%2F3727%2F7491%2Ffiles%2Ft-shirt-1.png%3Fv%3D1689798965&w=3840&q=75',
        category: 'Product',
      },
      {
        id: 1,
        title: 'Product Detail',
        price: 20,
        thumbnail:
          'https://demo.vercel.store/_next/image?url=https%3A%2F%2Fcdn.shopify.com%2Fs%2Ffiles%2F1%2F0754%2F3727%2F7491%2Ffiles%2Ft-shirt-1.png%3Fv%3D1689798965&w=3840&q=75',
        category: 'Product',
      },
    ],
    handleToggleCart: () => {},
  },
};
