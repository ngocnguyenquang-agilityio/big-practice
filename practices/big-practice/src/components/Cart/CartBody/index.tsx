// Stores
import { useCart } from '@stores/cart';

// Components
import ProductCartItem from '../ProductCartItem';
import CartEmpty from '../CartEmpty';

const CartBody = () => {
  const { cartItems } = useCart();

  const totalPrice = cartItems.reduce((total, item) => total + item.product.price * item.quantity, 0);

  return cartItems.length > 0 ? (
    <div className='flex h-full flex-col justify-between overflow-hidden p-1'>
      <ul className='flex-grow overflow-auto py-4'>
        {cartItems.map((item) => (
          <li
            key={item.id}
            className='flex w-full flex-col border-b border-neutral-700'
          >
            <ProductCartItem
              id={item.id}
              product={item.product}
              quantity={item.quantity}
            />
          </li>
        ))}
      </ul>
      <div className='py-4 text-sm text-neutral-400'>
        <div className='mb-3 flex items-center justify-between border-b pb-1 border-neutral-700'>
          <p>Taxes</p>
          <p className='text-right text-base text-white'>
            $0.00
            <span className='ml-1 inline'>USD</span>
          </p>
        </div>
        <div className='mb-3 flex items-center justify-between border-b pb-1 pt-1 border-neutral-700'>
          <p>Shipping</p>
          <p className='text-right'>Calculated at checkout</p>
        </div>
        <div className='mb-3 flex items-center justify-between border-b pb-1 pt-1 border-neutral-700'>
          <p>Total</p>
          <p className='text-right text-base text-white'>
            ${totalPrice}
            <span className='ml-1 inline'>USD</span>
          </p>
        </div>
      </div>
    </div>
  ) : (
    <CartEmpty />
  );
};

export default CartBody;
