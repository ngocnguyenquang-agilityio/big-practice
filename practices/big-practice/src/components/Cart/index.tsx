// Icons
import { CrossIcon } from '@assets/icons';

// Components
import { Button } from '@components/Button';
import CartEmpty from '@components/Cart/CartEmpty';
import CartBody from '@components/Cart/CartBody';

// Types
import { IProduct } from '@interfaces';

interface ICart {
  handleToggleCart: () => void;
  cart?: IProduct[];
}

const Cart = ({ handleToggleCart, cart = [] }: ICart) => {
  return (
    <aside className='fixed bottom-0 right-0 top-0 flex h-full w-full flex-col border-l p-6 backdrop-blur-xl border-neutral-700 bg-black/80 text-white md:w-[390px] translate-x-0 z-10'>
      <div className='flex items-center justify-between'>
        <p className='text-lg font-semibold'>My Cart</p>
        <Button
          data-testid='close-cart-btn'
          variant='outline'
          size='icon'
          onClick={handleToggleCart}
        >
          <CrossIcon />
        </Button>
      </div>
      {cart.length > 0 ? <CartBody /> : <CartEmpty />}
    </aside>
  );
};

export default Cart;
