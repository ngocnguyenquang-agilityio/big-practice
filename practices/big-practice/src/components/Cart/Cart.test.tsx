import { render, screen, fireEvent } from '@testing-library/react';

// Components
import Cart, { ICart } from '.';

// Mocks
import { mockCart } from '@mocks';

const props: ICart = {
  handleToggleCart: jest.fn(),
  cart: mockCart,
};

describe('Cart Component', () => {
  test('Should render successfully', () => {
    const comp = render(<Cart {...props} />);
    expect(comp).toBeTruthy();
  });

  test('Should match snapshot', () => {
    const comp = render(<Cart {...props} />);
    expect(comp).toMatchSnapshot();
  });

  test('Should render empty cart if no data', () => {
    const mockFunc = jest.fn();
    render(
      <Cart
        cart={[]}
        handleToggleCart={mockFunc}
      />,
    );
    const emptyCart = screen.queryByText('Your cart is empty.');
    expect(emptyCart).toBeTruthy();
  });

  test('Should call handleToggleCart', () => {
    render(<Cart {...props} />);
    const button = screen.getByTestId('close-cart-btn');
    fireEvent.click(button);
    expect(props.handleToggleCart).toBeCalledTimes(1);
  });
});
