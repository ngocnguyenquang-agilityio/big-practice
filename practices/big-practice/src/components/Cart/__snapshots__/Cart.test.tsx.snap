// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart Component Should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <aside
        class="fixed bottom-0 right-0 top-0 flex h-full w-full flex-col border-l p-4 lg:px-6 backdrop-blur-xl border-neutral-700 bg-black/80 text-white md:w-[390px] translate-x-0 z-10"
      >
        <div
          class="flex items-center justify-between"
        >
          <p
            class="text-lg font-semibold"
          >
            My Cart
          </p>
          <div
            class="flex justify-end md:w-1/3"
          >
            <button
              class="inline-flex items-center text-white justify-center text-sm font-medium focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 ring-offset-slate-950 focus-visible:ring-slate-300 border border-neutral-700 transition-colors h-11 w-11 rounded-md"
              data-testid="close-cart-btn"
            >
              <img
                alt="cross-icon"
                src=""
              />
            </button>
          </div>
        </div>
        <div
          class="flex h-full flex-col justify-between overflow-hidden p-1"
        >
          <ul
            class="flex-grow overflow-auto py-4"
          />
          <div
            class="py-4 text-sm text-neutral-400"
          >
            <div
              class="mb-3 flex items-center justify-between border-b pb-1 border-neutral-700"
            >
              <p>
                Taxes
              </p>
              <p
                class="text-right text-base text-white"
              >
                $0.00
                <span
                  class="ml-1 inline"
                >
                  USD
                </span>
              </p>
            </div>
            <div
              class="mb-3 flex items-center justify-between border-b pb-1 pt-1 border-neutral-700"
            >
              <p>
                Shipping
              </p>
              <p
                class="text-right"
              >
                Calculated at checkout
              </p>
            </div>
            <div
              class="mb-3 flex items-center justify-between border-b pb-1 pt-1 border-neutral-700"
            >
              <p>
                Total
              </p>
              <p
                class="text-right text-base text-white"
              >
                $
                0
                <span
                  class="ml-1 inline"
                >
                  USD
                </span>
              </p>
            </div>
          </div>
        </div>
      </aside>
    </div>
  </body>,
  "container": <div>
    <aside
      class="fixed bottom-0 right-0 top-0 flex h-full w-full flex-col border-l p-4 lg:px-6 backdrop-blur-xl border-neutral-700 bg-black/80 text-white md:w-[390px] translate-x-0 z-10"
    >
      <div
        class="flex items-center justify-between"
      >
        <p
          class="text-lg font-semibold"
        >
          My Cart
        </p>
        <div
          class="flex justify-end md:w-1/3"
        >
          <button
            class="inline-flex items-center text-white justify-center text-sm font-medium focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 ring-offset-slate-950 focus-visible:ring-slate-300 border border-neutral-700 transition-colors h-11 w-11 rounded-md"
            data-testid="close-cart-btn"
          >
            <img
              alt="cross-icon"
              src=""
            />
          </button>
        </div>
      </div>
      <div
        class="flex h-full flex-col justify-between overflow-hidden p-1"
      >
        <ul
          class="flex-grow overflow-auto py-4"
        />
        <div
          class="py-4 text-sm text-neutral-400"
        >
          <div
            class="mb-3 flex items-center justify-between border-b pb-1 border-neutral-700"
          >
            <p>
              Taxes
            </p>
            <p
              class="text-right text-base text-white"
            >
              $0.00
              <span
                class="ml-1 inline"
              >
                USD
              </span>
            </p>
          </div>
          <div
            class="mb-3 flex items-center justify-between border-b pb-1 pt-1 border-neutral-700"
          >
            <p>
              Shipping
            </p>
            <p
              class="text-right"
            >
              Calculated at checkout
            </p>
          </div>
          <div
            class="mb-3 flex items-center justify-between border-b pb-1 pt-1 border-neutral-700"
          >
            <p>
              Total
            </p>
            <p
              class="text-right text-base text-white"
            >
              $
              0
              <span
                class="ml-1 inline"
              >
                USD
              </span>
            </p>
          </div>
        </div>
      </div>
    </aside>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
